feat: configuring a user namespace (#2674)
codenvy/che-docs: modules/administration-guide/pages/configuring-a-user-namespace.adoc

:_content-type: PROCEDURE
:description: Configuring a user namespace
:keywords: administration guide, configuring, user, namespace
:navtitle: Configuring a user namespace
:page-aliases:

[id="configuring-a-user-namespace"]
= Configuring a user namespace

> This procedure walks through the process of using the {prod-short}

Reviewer: ```suggestion
This procedure walks you through the process of using {prod-short}
```

to replicate `ConfigMaps`, `Secrets` and `PersistentVolumeClaim` from `{prod-namespace}` namespace to numerous

> user-specific namespaces. The operator automates the synchronization of important configuration

Reviewer: Which operator is it? Does it have a name e.g. Dev Spaces Operator?

Author: It is `{prod}` operator

> data such as passwords, ssh-keys, certificates etc to users' namespaces.

Reviewer: ```suggestion
data such as passwords, SSH keys, and certificates to user namespaces.
```

> If a {kubernetes} resource is changed in an {prod-namespace} namespace,

Reviewer: ```suggestion
If you make changes to a {kubernetes} resource in an {prod-namespace} namespace,
```

the operator will immediately replicate the changes across all users namespaces.
In reverse, if a {kubernetes} resource is modified in a user namespace,
the operator will immediately revert the changes.

> The special logic is used for 'PersistentVolumeClaim' resources. If a 'PersistentVolumeClaim'
> is updated in {prod-namespace} or user namespaces, it will not be replicated or reverted due

Reviewer: ```suggestion
Special logic is used for 'PersistentVolumeClaim' resources. If you update 'PersistentVolumeClaim'
in {prod-namespace} or user namespaces, it will not be replicated or reverted due
```

> to its readonly specification. To modify the 'PersistentVolumeClaim' specification, delete it and create a new one

Reviewer: ```suggestion
to its read-only specification. To modify the 'PersistentVolumeClaim' specification, delete it and create a new one
```

See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
for other possible setup labels and annotations.
+
[source,yaml,subs="+attributes,+quotes"]
----
kind: Secret
apiVersion: v1
metadata:
  name: user-secret
  namespace: {prod-namespace}
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: workspaces-config
stringData:
...
----

. Create the `PersistentVolumeClaim` below to replicate it to every user namespace.
+
To enhance the configurability, you can customize the `PersistentVolumeClaim` by adding additional labels and annotations.
See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
for other possible setup labels and annotations.
+
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: user-pvc

>   namespace: eclipse-che

Author: ```suggestion
  namespace: {prod-namespace}
```

> See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
> for other possible setup labels and annotations.
> +
> [source,yaml,subs="+attributes,+quotes"]
> ----
> kind: ConfigMap
> apiVersion: v1
> metadata:
>   name: user-configmap
>   namespace: {prod-namespace}
>   labels:
>     app.kubernetes.io/part-of: che.eclipse.org
>     app.kubernetes.io/component: workspaces-config
> data:
> ...
> ----
> 
> * Create the `Secret` below to replicate it to every user namespace.
> +
> The following labels will be added to a replicated Secret automatically:
> `"controller.devfile.io/watch-secret":"true"` and `"controller.devfile.io/mount-to-devworkspace":"true"`.
> +
> To enhance the configurability, you can customize the `Secret` by adding additional labels and annotations.
> See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
> for other possible setup labels and annotations.
> +
> [source,yaml,subs="+attributes,+quotes"]
> ----
> kind: Secret
> apiVersion: v1
> metadata:
>   name: user-secret
>   namespace: {prod-namespace}
>   labels:
>     app.kubernetes.io/part-of: che.eclipse.org
>     app.kubernetes.io/component: workspaces-config
> stringData:
> ...
> ----
> 
> * Create the `PersistentVolumeClaim` below to replicate it to every user namespace.
> +
> To enhance the configurability, you can customize the `PersistentVolumeClaim` by adding additional labels and annotations.
> See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
> for other possible setup labels and annotations.
> +
> [source,yaml,subs="+attributes,+quotes"]
> ----
> apiVersion: v1
> kind: PersistentVolumeClaim
> metadata:
>   name: user-pvc
>   namespace: eclipse-che
>   labels:
>     app.kubernetes.io/part-of: che.eclipse.org
>     app.kubernetes.io/component: workspaces-config
> spec:
> ...
> ----
> 

Reviewer: ```suggestion
. Create the `ConfigMap` below to replicate it to every user namespace.
+
The following labels will be added to a replicated ConfigMap automatically:
`"controller.devfile.io/watch-configmap":"true"` and `"controller.devfile.io/mount-to-devworkspace":"true"`.
+
To enhance the configurability, you can customize the `Secret` by adding additional labels and annotations.
See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
for other possible setup labels and annotations.
+
[source,yaml,subs="+attributes,+quotes"]
----
kind: ConfigMap
apiVersion: v1
metadata:
  name: user-configmap
  namespace: {prod-namespace}
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: workspaces-config
data:
...
----

. Create the `Secret` below to replicate it to every user namespace.
+
The following labels will be added to a replicated Secret automatically:
`"controller.devfile.io/watch-secret":"true"` and `"controller.devfile.io/mount-to-devworkspace":"true"`.
+
To enhance the configurability, you can customize the `Secret` by adding additional labels and annotations.
See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
for other possible setup labels and annotations.
+
[source,yaml,subs="+attributes,+quotes"]
----
kind: Secret
apiVersion: v1
metadata:
  name: user-secret
  namespace: {prod-namespace}
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: workspaces-config
stringData:
...
----

. Create the `PersistentVolumeClaim` below to replicate it to every user namespace.
+
To enhance the configurability, you can customize the `PersistentVolumeClaim` by adding additional labels and annotations.
See the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets[Automatically mounting volumes, configmaps, and secrets]
for other possible setup labels and annotations.
+
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: user-pvc
  namespace: eclipse-che
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: workspaces-config
spec:
...
----

```
